File: sources/LiquidSense Release/me/aquavit/liquidsense/module/modules/render/BlockOverlay.java
package me.aquavit.liquidsense.module.modules.render;

import me.aquavit.liquidsense.event.EventTarget;
import me.aquavit.liquidsense.event.events.Render2DEvent;
import me.aquavit.liquidsense.event.events.Render3DEvent;
import me.aquavit.liquidsense.module.Module;
import me.aquavit.liquidsense.module.ModuleCategory;
import me.aquavit.liquidsense.module.ModuleInfo;
import me.aquavit.liquidsense.ui.font.Fonts;
import me.aquavit.liquidsense.utils.block.BlockUtils;
import me.aquavit.liquidsense.utils.render.ColorUtils;
import me.aquavit.liquidsense.utils.render.RenderUtils;
import me.aquavit.liquidsense.value.BoolValue;
import me.aquavit.liquidsense.value.IntegerValue;
import net.minecraft.block.Block;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderGlobal;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.BlockPos;
import org.lwjgl.opengl.GL11;

import java.awt.*;

@ModuleInfo(name = "BlockOverlay", description = "Allows you to change the design of the block overlay.", category = ModuleCategory.RENDER)
public class BlockOverlay extends Module {

    private IntegerValue colorRedValue = new IntegerValue("R", 68, 0, 255);
    private IntegerValue colorGreenValue = new IntegerValue("G", 117, 0, 255);
    private IntegerValue colorBlueValue = new IntegerValue("B", 255, 0, 255);
    private BoolValue colorRainbow = new BoolValue("Rainbow", false);

    public static BoolValue infoValue = new BoolValue("Info", false);

    public BlockPos getCurrentBlock() {
        BlockPos blockPos = mc.objectMouseOver.getBlockPos();
        if (blockPos == null)return null;

        if (BlockUtils.canBeClicked(blockPos) && mc.theWorld.getWorldBorder().contains(blockPos))
            return blockPos;

        return null;
    }

    @EventTarget
    public void onRender3D(Render3DEvent event) {
        BlockPos blockPos = getCurrentBlock();
        if (blockPos == null) return;
        Block block = mc.theWorld.getBlockState(blockPos).getBlock();
        if (block == null) return;
        float partialTicks = event.getPartialTicks();
        Color color = colorRainbow.get() ? ColorUtils.rainbow((long) 0.4f) : new Color(colorRedValue.get(), colorGreenValue.get(), colorBlueValue.get(), 102);

        GlStateManager.enableBlend();
        GlStateManager.tryBlendFuncSeparate(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA, GL11.GL_ONE, GL11.GL_ZERO);
        RenderUtils.glColor(color);
        GL11.glLineWidth(2F);
        GlStateManager.disableTexture2D();
        GlStateManager.depthMask(false);

        block.setBlockBoundsBasedOnState(mc.theWorld, blockPos);

        double x = mc.thePlayer.lastTickPosX + (mc.thePlayer.posX - mc.thePlayer.lastTickPosX) * partialTicks;
        double y = mc.thePlayer.lastTickPosY + (mc.thePlayer.posY - mc.thePlayer.lastTickPosY) * partialTicks;
        double z = mc.thePlayer.lastTickPosZ + (mc.thePlayer.posZ - mc.thePlayer.lastTickPosZ) * partialTicks;

        AxisAlignedBB axisAlignedBB = block.getSelectedBoundingBox(mc.theWorld, blockPos)
                .expand(0.0020000000949949026, 0.0020000000949949026, 0.0020000000949949026)
                .offset(-x, -y, -z);

        RenderGlobal.drawSelectionBoundingBox(axisAlignedBB);
        RenderUtils.drawFilledBox(axisAlignedBB);
        GlStateManager.depthMask(true);
        GlStateManager.enableTexture2D();
        GlStateManager.disableBlend();
        GlStateManager.resetColor();
    }

    @EventTarget
    public void onRender2D(Render2DEvent event) {
        if (infoValue.get()) {
            BlockPos blockPos = getCurrentBlock();
            if (blockPos == null) return;
            Block block = mc.theWorld.getBlockState(blockPos).getBlock();
            if (block == null) return;

            String info = block.getLocalizedName() + " §7ID: " + Block.getIdFromBlock(block);
            ScaledResolution scaledResolution = new ScaledResolution(mc);

            RenderUtils.drawBorderedRect(
                    (float) scaledResolution.getScaledWidth() / 2 - 2F,
                    (float) scaledResolution.getScaledHeight() / 2 + 5F,
                    (float) scaledResolution.getScaledWidth() / 2 + Fonts.font20.getStringWidth(info) + 2F,
                    (float) scaledResolution.getScaledHeight() / 2 + 16F,
                    3F, Color.BLACK.getRGB(), Color.BLACK.getRGB()
            );
            GlStateManager.resetColor();
            Fonts.font20.drawString(info, scaledResolution.getScaledWidth() / 2, scaledResolution.getScaledHeight() / 2 + 7,
                    Color.WHITE.getRGB());
        }
    }
}

File: sources/Kevin Reborn/kevin/module/modules/render/BlockOverlay.kt
/*
 * This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package kevin.module.modules.render

import kevin.event.EventTarget
import kevin.event.Render3DEvent
import kevin.module.*
import kevin.utils.BlockUtils.canBeClicked
import kevin.utils.ColorUtils.rainbow
import kevin.utils.RenderUtils
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.util.BlockPos
import org.lwjgl.opengl.GL11
import java.awt.Color

class BlockOverlay : Module("BlockOverlay", "Allows you to change the design of the block overlay.", category = ModuleCategory.RENDER) {
    private val colorRedValue = IntegerValue("R", 68, 0, 255)
    private val colorGreenValue = IntegerValue("G", 117, 0, 255)
    private val colorBlueValue = IntegerValue("B", 255, 0, 255)
    private val lineWidth = FloatValue("LineWidth",5F,2F,10F)
    private val colorRainbow = BooleanValue("Rainbow", false)
    private val onlyOutlineValue = BooleanValue("OnlyOutline", false)

    val currentBlock: BlockPos?
        get() {
            val blockPos = mc.objectMouseOver?.blockPos ?: return null

            if (canBeClicked(blockPos) && mc.theWorld!!.worldBorder.contains(blockPos))
                return blockPos

            return null
        }

    @EventTarget
    fun onRender3D(event: Render3DEvent) {
        val blockPos = currentBlock ?: return

        val block = mc.theWorld!!.getBlockState(blockPos).block
        val partialTicks = event.partialTicks

        val color = if (colorRainbow.get()) rainbow(0.4F) else Color(colorRedValue.get(),
            colorGreenValue.get(), colorBlueValue.get(), (0.4F * 255).toInt())

        GlStateManager.enableBlend()
        GlStateManager.tryBlendFuncSeparate(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA, GL11.GL_ONE, GL11.GL_ZERO)
        RenderUtils.glColor(color)
        GL11.glLineWidth(2F)
        GlStateManager.disableTexture2D()
        GL11.glDepthMask(false)

        block.setBlockBoundsBasedOnState(mc.theWorld!!, blockPos)

        val thePlayer = mc.thePlayer ?: return

        val x = thePlayer.lastTickPosX + (thePlayer.posX - thePlayer.lastTickPosX) * partialTicks
        val y = thePlayer.lastTickPosY + (thePlayer.posY - thePlayer.lastTickPosY) * partialTicks
        val z = thePlayer.lastTickPosZ + (thePlayer.posZ - thePlayer.lastTickPosZ) * partialTicks

        val axisAlignedBB = block.getSelectedBoundingBox(mc.theWorld!!, blockPos)
            .expand(0.0020000000949949026, 0.0020000000949949026, 0.0020000000949949026)
            .offset(-x, -y, -z)

        GL11.glLineWidth(lineWidth.get())
        RenderUtils.drawSelectionBoundingBox(axisAlignedBB)
        if (!onlyOutlineValue.get()) RenderUtils.drawFilledBox(axisAlignedBB)
        GL11.glDepthMask(true)
        GlStateManager.enableTexture2D()
        GlStateManager.disableBlend()
        GlStateManager.resetColor()
    }
}